File: 200-number-of-islands.js
// Given a 2d grid map of '1's (land) and '0's (water), count the number
// of islands. An island is surrounded by water and is formed by
// connecting adjacent lands horizontally or vertically. You may assume
// all four edges of the grid are all surrounded by water.

// Example 1:

// Input:
// 11110
// 11010
// 11000
// 00000

// Output: 1
// Example 2:

// Input:
// 11000
// 11000
// 00100
// 00011

// Output: 3


// This solution isn't optimal, it is a trial of union-find algorithm
// using.

class UnionFindCompressed {
    /**
     * Create a parents data array.
     * @param {number} n
     */
    constructor(n) {
        this.parents = new Array(n);
        for (let i = 0; i < n; i++) {
            this.parents[i] = i;
        }
    }

    /**
     * Find a node root.
     * @param {number} index
     */
    root(index) {
        while (this.parents[index] !== index) {
            this.parents[index] = this.parents[this.parents[index]];
            index = this.parents[index];
        }
        return index;
    }

    /**
     * Are the nodes connected?
     *
     * @param {number} a
     * @param {number} b
     *
     * @returns {boolean} Check result.
     */
    connected(a, b) {
        return this.root(a) === this.root(b);
    }

    /**
     * Connect two nodes.
     *
     * @param {number} a
     * @param {number} b
     */
    union(a, b) {
        const ra = this.root(a);
        const rb = this.root(b);
        this.parents[ra] = rb;
    }
}

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    const rows = grid.length;
    if (rows === 0) {
        return 0;
    }
    const cols = grid[0].length;

    const uf = new UnionFindCompressed(rows * cols);
    toFlatIndex = (row, col) => row * cols + col;
    const land = new Set();

    for (let row = 0; row < rows; row++) {
        for (let col = 0; col < cols; col++) {
            if (grid[row][col] === '0') {
                continue;
            }
            const currentIndex = toFlatIndex(row, col);
            land.add(currentIndex);

            // Check vertical connectivity.
            if (row > 0 && grid[row - 1][col] === '1') {
                uf.union(currentIndex, toFlatIndex(row - 1, col));
            }
            if (row < rows - 1 && grid[row + 1][col] === '1') {
                uf.union(currentIndex, toFlatIndex(row + 1, col));
            }

            // Check horizontal connectivity.
            if (col > 0 && grid[row][col - 1] === '1') {
                uf.union(currentIndex, toFlatIndex(row, col - 1));
            }
            if (col < cols - 1 && grid[row][col + 1] === '1') {
                uf.union(currentIndex, toFlatIndex(row, col + 1));
            }
        }
    }

    const roots = new Set();
    for (let item of land) {
        roots.add(uf.root(item));
    }
    return roots.size;
};

const t1 = ["11110", "11010", "11000", "00000"];
console.log("1 = ", numIslands(t1));

const t2 = ["11000", "11000", "00100", "00011"];
console.log("3 = ", numIslands(t2));

const t3 = [];
console.log("0 = ", numIslands(t3));

const t4 = ["1"];
console.log("1 = ", numIslands(t4));

const t5 = ["0"];
console.log("0 = ", numIslands(t5));

const t6 = [["1","0","1","1","0","1","1"]];
console.log("3 = ", numIslands(t6));

const t7 = [["1"],["0"],["1"],["1"],["0"],["1"],["1"]];
console.log("3 = ", numIslands(t7));

File: 200-number-of-islands.py
# Number of Islands

# Given a 2d grid map of '1's (land) and '0's (water), count the number
# of islands. An island is surrounded by water and is formed by
# connecting adjacent lands horizontally or vertically. You may assume
# all four edges of the grid are all surrounded by water.

# Example 1:
# Input:
# 11110
# 11010
# 11000
# 00000
# Output: 1

# Example 2:
# Input:
# 11000
# 11000
# 00100
# 00011
# Output: 3

from typing import List


class Solution:
    def sink_island(self, row, col, rc, cc, grid):
        grid[row][col] = "0"
        if row > 0 and grid[row - 1][col] == "1":
            self.sink_island(row - 1, col, rc, cc, grid)
        if row < rc - 1 and grid[row + 1][col] == "1":
            self.sink_island(row + 1, col, rc, cc, grid)
        if col > 0 and grid[row][col - 1] == "1":
            self.sink_island(row, col - 1, rc, cc, grid)
        if col < cc - 1 and grid[row][col + 1] == "1":
            self.sink_island(row, col + 1, rc, cc, grid)

    def numIslands(self, grid: List[List[str]]) -> int:
        islands = 0
        rc = len(grid)
        if not rc: return 0
        cc = len(grid[0])
        if not cc: return 0

        for i in range(0, rc):
            for j in range(0, cc):
                if grid[i][j] == "1":
                    islands +=1
                    self.sink_island(i, j, rc, cc, grid)
        return islands

t = Solution()

input = [["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]
print("1 = ", t.numIslands(input))

input = [["1","1","0","0","0"],["1","1","0","0","0"],["0","0","1","0","0"],["0","0","0","1","1"]]
print("3 = ", t.numIslands(input))

input = [["0","0","0","0","0"],["0","0","0","0","0"]]
print("0 = ", t.numIslands(input))

input = [["1","1","0","1","0"]]
print("2 = ", t.numIslands(input))

input = [["1"],["1"],["0"],["1"],["0"]]
print("2 = ", t.numIslands(input))

input = [[]]
print("0 = ", t.numIslands(input))

input = []
print("0 = ", t.numIslands(input))
